sbc_compassion/models/import_letters_history.py

```diff
@@ -233,57 +155,62 @@ def onchange_config(self):
             for field, val in list(config.get_correspondence_metadata().items()):
                 setattr(self, field, val)
 
-    ##########################################################################
-    #                             PRIVATE METHODS                            #
-    ##########################################################################
+    def local_generator(self):
+        d = set(self.data)
+        n = len(d)
+        for i, attachment in enumerate(d):
+            yield i + 1, n, attachment.name
+            pdf_data = base64.b64decode(attachment.with_context(bin_size=False).datas)
+            self._analyze_pdf(pdf_data, attachment.name)
+
     @job(default_channel="root.sbc_compassion")
     @related_action(action="related_action_s2b_imports")
-    def _run_analyze(self):
-        """
-        Analyze each attachment:
-        - check for duplicate file names and skip them
-        - decompress zip file if necessary
-        - call _analyze_attachment for every resulting file
-        """
+    def run_analyze(self, generator=None):
+        if generator is None:
+            generator = self.local_generator
+
         self.ensure_one()
-        # keep track of file names to detect duplicates
-        file_name_history = []
-        logger.info("Imported files analysis started...")
-        progress = 1
+        self.state = "pending"
+        logger.info("Letters import started...")
 
-        for attachment in self.data:
-            if attachment.name not in file_name_history:
-                file_name_history.append(attachment.name)
-                file_data = base64.b64decode(
-                    attachment.with_context(bin_size=False).datas
-                )
-                # check for zip
-                if func.check_file(attachment.name) == 2:
-                    zip_file = BytesIO(file_data)
-                    zip_ = zipfile.ZipFile(zip_file, "r")
-                    for f in zip_.namelist():
-                        logger.debug(f"Analyzing file {progress}/{self.nber_letters}")
-                        self._analyze_attachment(zip_.read(f), f)
-                        progress += 1
-                # case with normal format (PDF,TIFF)
-                elif func.check_file(attachment.name) == 1:
-                    logger.debug(f"Analyzing file {progress}/{self.nber_letters}")
-                    self._analyze_attachment(file_data, attachment.name)
-                    progress += 1
-                else:
-                    raise UserError(_("Only zip/pdf files are supported."))
-            else:
-                raise UserError(_("Two files are the same"))
+        for i, n, file in generator():
```

Comment:
Please use human readable variable names. Single characters are not a good practice and make it difficult to understand the code.